library/alloc/src/boxed/thin.rs

```diff
@@ -91,6 +93,88 @@ impl<T> ThinBox<T> {
 
 #[unstable(feature = "thin_box", issue = "92791")]
 impl<Dyn: ?Sized> ThinBox<Dyn> {
+    #[cfg(not(no_global_oom_handling))]
+    fn new_unsize_zst<T>(value: T) -> Self
+    where
+        T: Unsize<Dyn>,
+    {
+        assert!(mem::size_of::<T>() == 0);
+
+        #[repr(C)]
+        struct ReprC<A, B> {
+            a: A,
+            b: B,
+        }
+
+        struct EmptyArray<T> {
+            _array: [T; 0],
+        }
+
+        // FIXME: empty array should be automatically `Freeze` for empty arrays.
```

Comment:
It's far from clear that that is what we want. There's tons of prior discussion here, e.g. https://github.com/rust-lang/unsafe-code-guidelines/issues/236.

Reply:
Removed FIXME, kept `Freeze`.